packages/fx-core/src/plugins/resource/cicd/plugin.ts

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import { Context } from "@microsoft/teamsfx-api/build/v2";
+import { Inputs, v2, Result, FxError, Platform } from "@microsoft/teamsfx-api";
+import { FxCICDPluginResultFactory as ResultFactory } from "./result";
+import { CICDProviderFactory } from "./providers/factory";
+import { ProviderKind } from "./providers/enums";
+import { questionNames } from "./questions";
+import { InternalError, NoProjectOpenedError } from "./errors";
+import { generateBuildScript } from "./utils/buildScripts";
+import { telemetryHelper } from "./utils/telemetry-helper";
+
+export class CICDImpl {
+  public async addCICDWorkflows(
+    context: Context,
+    inputs: Inputs,
+    envInfo: v2.EnvInfoV2
+  ): Promise<Result<any, FxError>> {
+    // 1. Key inputs (envName, provider, template) x (hostingType, ).
+    if (!inputs.projectPath) {
+      throw new NoProjectOpenedError();
```

Comment:
It looks like only check if projectPath is defined, why we throw error as "No project opened"? It make sense that throw internal error that project path is not defined?

Reply:
I believe if projectPath is undefined, it has a good chance users didn't open any workspace in VSCode, so this error makes sense. This follows the existing pattern: https://github.com/OfficeDev/TeamsFx/blob/306d41b05a4b73519374ed836fefb74fe7bf95bf/packages/fx-core/src/core/middleware/envInfoLoader.ts#L70